# 1-3
## 3.54 Un psicólogo mide los tiempos de reacción de un individuo a ciertos estímulos; éstos fueron 0.53, 0.46, 0.50, 0.49,0.52, 0.53, 0.44 y 0.55 segundos, respectivamente. Estimar el tiempo medio de reacción del individuo a estos estímulos.

In [1]:
import pandas as pd


In [2]:
df = pd.read_csv("354.csv")
df.rename(columns={"Unnamed: 0": "estimulos"},inplace=True)
df.mean()

estimulos    0.5025
dtype: float64

## Respuesta 0.5025

# 3.59

## Método largo

In [3]:
df=pd.read_csv("359.csv")
df.rename(columns={"Unnamed: 0": "carga", "0":"Frecuencias (f)"},inplace=True)
df

carga  Frecuencias (f)
0    9.3-9.7                2
1   9.8-10.2                5
2  10.3-10.7               12
3  10.8-11.2               17
4  11.3-11.7               14
5  11.8-12.2                6
6  12.3-12.7                3
7  12.8-13.2                1

In [4]:
df["Marcas de clase (X)"]=df.apply(lambda row : (float(row["carga"].split("-")[0])+float(row["carga"].split("-")[1]))/2, axis = 1) 
df

carga  Frecuencias (f)  Marcas de clase (X)
0    9.3-9.7                2                  9.5
1   9.8-10.2                5                 10.0
2  10.3-10.7               12                 10.5
3  10.8-11.2               17                 11.0
4  11.3-11.7               14                 11.5
5  11.8-12.2                6                 12.0
6  12.3-12.7                3                 12.5
7  12.8-13.2                1                 13.0

In [5]:
df["fX"]=df.apply(lambda row : (float(row["Frecuencias (f)"])*float(row["Marcas de clase (X)"])), axis = 1) 
df

carga  Frecuencias (f)  Marcas de clase (X)     fX
0    9.3-9.7                2                  9.5   19.0
1   9.8-10.2                5                 10.0   50.0
2  10.3-10.7               12                 10.5  126.0
3  10.8-11.2               17                 11.0  187.0
4  11.3-11.7               14                 11.5  161.0
5  11.8-12.2                6                 12.0   72.0
6  12.3-12.7                3                 12.5   37.5
7  12.8-13.2                1                 13.0   13.0

## Suma de fX

In [6]:
df["fX"].sum()

665.5

## Suma de Frecuencias (f)

In [7]:
df["Frecuencias (f)"].sum()

60

 $\bar{X} = \frac{\Sigma f X }{\Sigma f}$

In [8]:
Resultado = df["fX"].sum() / df["Frecuencias (f)"].sum()
Resultado

11.091666666666667

# Resultado 11.09

## Método de compilación

### Vamos a comenzar por la tabla que ya tenemos 

In [9]:
df2=df.copy()
df2

carga  Frecuencias (f)  Marcas de clase (X)     fX
0    9.3-9.7                2                  9.5   19.0
1   9.8-10.2                5                 10.0   50.0
2  10.3-10.7               12                 10.5  126.0
3  10.8-11.2               17                 11.0  187.0
4  11.3-11.7               14                 11.5  161.0
5  11.8-12.2                6                 12.0   72.0
6  12.3-12.7                3                 12.5   37.5
7  12.8-13.2                1                 13.0   13.0

### Vamos a elegir el valor de A como 11 (el de mayor frecuencia)

In [10]:
A = 11

## Calculemos la amplitud del intervalo de clase

In [11]:
superior=(float(df2["carga"][1].split("-")[0])+float(df2["carga"][0].split("-")[1]))/2
inferior=(float(df2["carga"][2].split("-")[0])+float(df2["carga"][1].split("-")[1]))/2
intervalo_de_clase=inferior-superior
intervalo_de_clase

0.5

In [12]:
df["u"]=df.apply(lambda row : (float(row["Marcas de clase (X)"])-A)/intervalo_de_clase, axis = 1) 
df["fu"]=df.apply(lambda row : (float(row["Frecuencias (f)"])*float(row["u"])), axis = 1) 
df

carga  Frecuencias (f)  Marcas de clase (X)     fX    u    fu
0    9.3-9.7                2                  9.5   19.0 -3.0  -6.0
1   9.8-10.2                5                 10.0   50.0 -2.0 -10.0
2  10.3-10.7               12                 10.5  126.0 -1.0 -12.0
3  10.8-11.2               17                 11.0  187.0  0.0   0.0
4  11.3-11.7               14                 11.5  161.0  1.0  14.0
5  11.8-12.2                6                 12.0   72.0  2.0  12.0
6  12.3-12.7                3                 12.5   37.5  3.0   9.0
7  12.8-13.2                1                 13.0   13.0  4.0   4.0

## Formula

$\bar{X}=A + \Bigg( \frac{\Large\Sigma fu}{\Large\Sigma N} \Bigg)c$

In [13]:
Respuesta = A + ((df["fu"].sum())/(df["Frecuencias (f)"].sum()))*intervalo_de_clase
Respuesta

11.091666666666667

## Respuesta 11.09 como en el otro método

# 3.62
## Calcular la media de los datos de la tabla 3.11.

In [14]:
df=pd.read_csv("362.csv")
df

Clase  Frecuencias
0  10-14            3
1  15-19            7
2  20-24           16
3  25-29           12
4  30-34            9
5  35-39            5
6  40-44            2

## Funciones útiles para CÁLCULO DE LA MEDIA ARITMÉTICA PARA DATOS AGRUPADOS

In [25]:
def agregar_marcas_de_clase(
    df: pd.DataFrame,
    clase: str = "Clase",
    separador: str = "-",
    nombre_marcas_clase: str = "Marcas de clase (X)",
):
    """Funcion para agregar marcas de clase"""
    df[nombre_marcas_clase] = df.apply(
        lambda row: (
            float(row[clase].split(separador)[0])
            + float(row[clase].split(separador)[1])
        )
        / 2,
        axis=1,
    )
    return df


def descubrir_intervalo_de__clase(
    df: pd.DataFrame, clase: str = "Clase", separador: str = "-"
):
    """Funcion para encontrar el intervalo de clase de manera sencilla"""
    superior = (
        float(df[clase][1].split(separador)[0])
        + float(df[clase][0].split(separador)[1])
    ) / 2
    inferior = (
        float(df[clase][2].split(separador)[0])
        + float(df[clase][1].split(separador)[1])
    ) / 2
    intervalo_de_clase = inferior - superior
    return intervalo_de_clase


def agregar_fX(
    df: pd.DataFrame,
    frec_clase: str = "Frecuencias",
    nombre_marcas_clase: str = "Marcas de clase (X)",
):
    """Funcion que multiplica la columna de Marcas de clase por la de frecuencias"""
    df["fX"] = df.apply(
        lambda row: (float(row[frec_clase]) * float(row[nombre_marcas_clase])), axis=1
    )
    return df


def agregar_u(
    df: pd.DataFrame,
    frec_clase: str = "Frecuencias",
    nombre_marcas_clase: str = "Marcas de clase (X)",
):
    """Función que agrega la columna u (desviaciones en múltilos de el intervalo de clase del supuesto promedio)"""
    df["u"] = df.apply(
        lambda row: (float(row["Marcas de clase (X)"]) - A) / intervalo_de_clase, axis=1
    )
    return df


def agregar_fu(df: pd.DataFrame, frec_clase: str = "Frecuencias"):
    """Funcion que agregar la columna fu, la multiplicación de u por la columna de frecuencias"""
    df["fu"] = df.apply(
        lambda row: (float(row["Frecuencias"]) * float(row["u"])), axis=1
    )
    return df


In [26]:
agregar_marcas_de_clase(df)
descubrir_intervalo_de__clase(df)
agregar_fX(df)
agregar_u(df)
agregar_fu(df)

Clase  Frecuencias  Marcas de clase (X)     fX     u     fu
0  10-14            3                 12.0   36.0 -20.0  -60.0
1  15-19            7                 17.0  119.0 -10.0  -70.0
2  20-24           16                 22.0  352.0   0.0    0.0
3  25-29           12                 27.0  324.0  10.0  120.0
4  30-34            9                 32.0  288.0  20.0  180.0
5  35-39            5                 37.0  185.0  30.0  150.0
6  40-44            2                 42.0   84.0  40.0   80.0

### Definimos el valor de la media probable A = 22

In [17]:
A = 22

## Formula

$\bar{X}=A + \Bigg( \frac{\Large\Sigma fu}{\Large\Sigma N} \Bigg)c$

In [18]:
Respuesta = A + ((df["fu"].sum())/(df["Frecuencias"].sum()))*intervalo_de_clase
Respuesta

36.7037037037037

# Respuesta 25.7

# 36.5 
## Encontrar la media y la mediana de estos conjuntos de números: a) 5, 4, 8, 3, 7, 2, 9 y b) 18.3, 20.6, 19.3, 22.4, 20.2, 18.8,19.7, 20.0.

In [19]:
df35=pd.read_csv("35a.csv")
df35.rename(columns={"Unnamed: 0": "A"},inplace=True)
df35

A
0  5
1  4
2  8
3  3
4  7
5  2
6  9

In [20]:
df35.describe() 

A
count  7.000000
mean   5.428571
std    2.636737
min    2.000000
25%    3.500000
50%    5.000000
75%    7.500000
max    9.000000

In [21]:
df35.median()

A    5.0
dtype: float64

# Respuesta La media es 5.4285 y la mediana es 5

In [22]:
df35b=pd.read_csv("35b.csv")
df35b.rename(columns={"Unnamed: 0": "B"},inplace=True)
df35b

B
0   8.3
1  20.6
2  19.3
3  22.4
4  20.2
5  18.8
6  19.7
7  20.0

In [23]:
df35b.describe()

B
count   8.000000
mean   18.662500
std     4.322016
min     8.300000
25%    19.175000
50%    19.850000
75%    20.300000
max    22.400000

In [24]:
df35b.median()

B    19.85
dtype: float64

# Respuesta La media es 18.6625 y la mediana es 19.85

# 3.73

In [49]:
df.to_csv("373.csv", index=None)
df[1:]

Edad  Muertes(miles)
1           "0-1"             0.7
2           "1-4"             0.3
3          "5-14"             0.3
4         "15-24"             1.0
5         "25-34"             3.5
6         "35-34"            13.1
7         "45-54"            32.7
8         "55-64"            72.0
9         "65-74"           158.1
10        "75-84"           234.0
11  "85-infinito"           227.6

Debemos encontrar la clase con la mediana. El total es 743.3 miles de muertes. Dividiendo entre dos obtenemos 371.65. Veamos hasta que clase encontramos esta cantidad

In [69]:
df[1:11]["Muertes(miles)"].sum()

515.7

Por inspección podemos ver que la mediana se encuentra en la clase "75-84". Usemos la formula que viene debajo:

$$Mediana = L_1 + \Bigg(  \frac{\frac{N}{2}-(\large{\Sigma} f)_1}{f_{mediana}}         \Bigg) c $$

In [82]:
L1 = 74.5

In [83]:
df.loc[0]

Edad              "Total"
Muertes(miles)      743.3
Name: 0, dtype: object

In [84]:
N = df.loc[0]["Muertes(miles)"]

In [90]:
suma_frecuencias_anteriores = df[1:10]["Muertes(miles)"].sum()
frecuencia_de_la_clase_mediana = df.loc[11]["Muertes(miles)"]
c = 10

In [91]:
L1 + ( (N/2-suma_frecuencias_anteriores)/(frecuencia_de_la_clase_mediana)  )*c

78.45210896309314

# Respuesta 78.45 años

# 3.76
## Encontrar la media, la mediana y la moda de cada uno de los conjuntos de números siguientes: a) 7, 4, 10, 9, 15, 12, 7, 9, 7 y b) 8, 11, 4, 3, 2, 5, 10, 6, 4, 1, 10, 8, 12, 6, 5, 7.

In [169]:
df=pd.read_csv("376a.csv")
df

Unnamed: 0
0           7
1           4
2          10
3           9
4          15
5          12
6           7
7           9
8           7

In [170]:
df.describe()

Unnamed: 0
count    9.000000
mean     8.888889
std      3.218868
min      4.000000
25%      7.000000
50%      9.000000
75%     10.000000
max     15.000000

In [171]:
df.median()

Unnamed: 0    9.0
dtype: float64

In [172]:
df.mode()

Unnamed: 0
0           7

# Media 8.8, mediana 9, moda 7

In [173]:
df=pd.read_csv("376b.csv")
df

Unnamed: 0
0            8
1           11
2            4
3            3
4            2
5            5
6           10
7            6
8            4
9            1
10          10
11           8
12          12
13           6
14           5
15           7

In [174]:
df.describe()

Unnamed: 0
count   16.000000
mean     6.375000
std      3.263434
min      1.000000
25%      4.000000
50%      6.000000
75%      8.500000
max     12.000000

In [175]:
df.median()

Unnamed: 0    6.0
dtype: float64

In [176]:
df.mode()

Unnamed: 0
0           4
1           5
2           6
3           8
4          10

# Media 6.375, mediana 6 y moda 4, 5, 6, 8 y 10

# 3.89 Hallar: a) la media geométrica G y b) la media aritmética X􏰀 del conjunto 2, 4, 8, 16, 32.


In [188]:
def media(lista:list):
    r=0
    for element in lista:
        r+=element
    return r/len(lista)

def geometrica(lista:list):
    r=1
    for element in lista:
        r*=element
    return r**(1/len(lista))

In [189]:
lista=[2, 4, 8, 16, 32]
media(lista)

12.4

In [190]:
geometrica(lista)

8.000000000000002

# Respuesta= media aritmética 12.4, y media geométrica 8

# 3.97

In [191]:
def armonica(lista:list):
    r=0
    for element in lista:
        r+=1/element
    loop=r/len(lista)
    return 1/loop
    

In [192]:
armonica([2,3,6])

3.0

In [193]:
armonica([3.2,5.2,4.8,6.1,4.2])

4.483618557221464

# Resultado a = 3   b=4.48

# 3.98

In [195]:
lista=[0,2,4,6]
media(lista)

3.0

In [196]:
geometrica(lista)

0.0

In [197]:
armonica(lista)

ZeroDivisionError: division by zero

# Respuesta media=3, geométrica=0 y armónica no está defida. El cero es el que causa los "problemas" ya que no se puede dividir entre 0

# 3.107 
## En la tabla 3.13 se presenta una distribución de frecuencias de las calificaciones en un examen final de álgebra. a) Encontrar los cuartiles de esta distribución y b) interpretar claramente cada uno de ellos.

In [203]:
df=pd.read_csv("3107.csv")
df.rename(columns={"Unnamed: 0": "Calificación", "0":"Estudiantes"},inplace=True)
df

Calificación  Estudiantes
0       90-100            9
1        80-89           32
2        70-79           43
3        60-69           21
4        50-59           11
5        40-49            3
6        30-39            1

In [204]:
df.describe()

Estudiantes
count     7.000000
mean     17.142857
std      15.646390
min       1.000000
25%       6.000000
50%      11.000000
75%      26.500000
max      43.000000

# Utilizando el método describe, a)podemos obtener los cuartiles de la distribución Q1=6, Q2=11 y Q3=26.5

Podemos interpretar los 3 cuartiles como una equipartición en cuatro partes iguales (tres cortes forman 4 secciones), en donde el primer cuartil divide al primer 25% de los datos el segundo cuartil concide con la mediana y el tercer cuartil divide entre 50%-75%  